v2/internal/reconcilers/entra/entra_securitygroup_reconciler.go

```diff
@@ -0,0 +1,499 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package entra
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+
+	"github.com/go-logr/logr"
+	msgraphsdkgo "github.com/microsoftgraph/msgraph-sdk-go"
+	"github.com/microsoftgraph/msgraph-sdk-go/groups"
+	msgraphmodels "github.com/microsoftgraph/msgraph-sdk-go/models"
+	"github.com/microsoftgraph/msgraph-sdk-go/models/odataerrors"
+	"github.com/rotisserie/eris"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	asoentra "github.com/Azure/azure-service-operator/v2/api/entra/v1"
+	"github.com/Azure/azure-service-operator/v2/internal/config"
+	"github.com/Azure/azure-service-operator/v2/internal/identity"
+	"github.com/Azure/azure-service-operator/v2/internal/reconcilers"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/conditions"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/secrets"
+)
+
+// EntraSecurityGroupReconciler reconciles an Entra security group.
+// TODO: Factor out common code shared with other Entra resources into entraGenericReconciler
+type EntraSecurityGroupReconciler struct {
+	reconcilers.ReconcilerCommon
+	ResourceResolver   *resolver.Resolver
+	CredentialProvider identity.CredentialProvider
+	Config             config.Values
+	EntraClientFactory EntraConnectionFactory
+}
+
+var _ genruntime.Reconciler = &EntraSecurityGroupReconciler{}
+
+func NewEntraSecurityGroupReconciler(
+	kubeClient kubeclient.Client,
+	entraClientFactory EntraConnectionFactory,
+	resourceResolver *resolver.Resolver,
+	positiveConditions *conditions.PositiveConditionBuilder,
+	cfg config.Values,
+) *EntraSecurityGroupReconciler {
+	return &EntraSecurityGroupReconciler{
+		ResourceResolver:   resourceResolver,
+		Config:             cfg,
+		EntraClientFactory: entraClientFactory,
+		ReconcilerCommon: reconcilers.ReconcilerCommon{
+			KubeClient:         kubeClient,
+			PositiveConditions: positiveConditions,
+		},
+	}
+}
+
+func (r *EntraSecurityGroupReconciler) CreateOrUpdate(
+	ctx context.Context,
+	log logr.Logger,
+	eventRecorder record.EventRecorder,
+	obj genruntime.MetaObject,
+) (ctrl.Result, error) {
+	group, err := r.asSecurityGroup(obj)
+	if err != nil {
+		return ctrl.Result{}, eris.Wrapf(err, "creating or updating security group %s", group.Name)
+	}
+
+	// TODO: Respect annotation serviceoperator.azure.com/reconcile-policy
+
+	// If we already know the Entra ID of the group (captured in an annotation), we can update it directly
+	if id, ok := getEntraID(obj); ok {
```

Comment:
minor, style: Could you refactor this to have one section which is:
// get the ID (either from `getEntraID` or `tryAdopt`), and then a single `r.update` call with that id?

Ignore this if that refactoring sucks for some other reason, the reason I ask is it just means that there's one path to update and one path to create, which is nice from a readability perspective.

Reply:
I originally had something along those lines but it was a bit convoluted handling all the edge cases correctly (especially around optional adoption); this refactoring was my attempt to make the flow easy to understand.